core/src/main/scala/org/apache/spark/util/collection/Utils.scala

```diff
@@ -42,6 +42,23 @@ private[spark] object Utils extends SparkCollectionUtils {
     ordering.leastOf(input.asJava, num).iterator.asScala
   }
 
+  /**
+   * Returns the last K elements from the input.
+   */
+  def takeLast[T](input: Iterator[T], num: Int): Iterator[T] = {
+    assert(num >= 0)
+    if (input.isEmpty || num == 0) {
+      Iterator.empty[T]
+    } else {
+      var last = Seq.empty[T]
+      val grouped = input.grouped(num)
+      while (grouped.hasNext) {
+        last = grouped.next()
+      }
+      last.iterator
+    }
```

Comment:
It seems that there is an error in this logic. `takeLastWith(Range(0, 100).iterator, 3).toSeq` returns `List(99)` instead of `List(97, 98, 99)`.

Perhaps a Queue should be used to temporarily store the result data？

```scala
assert(num >= 0)
    if (input.isEmpty || num == 0) {
      Iterator.empty[T]
    } else {
      val queue = mutable.Queue.empty[T]
      input.foreach { item =>
        if (queue.size >= num) {
          queue.dequeue()
        }
        queue.enqueue(item)
      }
      queue.iterator
    }

```

Reply:
yes, it should be `sliding`

Reply:
Wait a moment, let me check the performance of `sliding`

Reply:
spark has a `HybridRowQueue`, let me take a look at it first